### PR TITLE
optimize current deposition `ZigZag`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- remove complicated min/max construct to calculate the relay
  point (paper conform change)
- not split the trajectory for a particle which not leaf the cell (optimization)
- return current deposition for a direction if flux is zero (optimization)
- remove particle splitting in 2D for the z component of the current (optimization)

Test:
- [x] heating 2D KHI positron/electron with z-drift
- [x] heating 2D KHI positron/electron with x-drift
- [x] heating 3D KHI positron/electron
- [x] speed tests

**Why it is not implemented as separate solver**
The difference to the paper implementation is very small and a second implementation results in 99% code duplication.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
heating 2D KHI positron/electron with z-drift (ZigZag-dev vs. current pull request)
- 16 GPUs k20
- 2000 steps

2d-z_heating

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 1]
```diff
@@ -307,6 +311,11 @@ struct ZigZag
 
     /** calculate virtual point were we split our particle trajectory
      *
+     * The relay point calculation differ from the paper version in the point
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: differs

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 2]
```diff
@@ -307,6 +311,11 @@ struct ZigZag
...
+     * that the trajectory of a particle which not leaf the cell is not split.
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: which do notlneave the is not splitted.

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 3]
```diff
@@ -307,6 +311,11 @@ struct ZigZag
...
+     * The relay point for a particle which not leaf the cell is set to the
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: does not leave

**Comment by psychocoderHPC** [comment]
heating 3D KHI positron/electron (ZigZag-dev vs. current pull request)

16 GPUs k20
2000 steps

3d-heating

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: are you sure that "relay point" is the correct term?

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 4]
```diff
@@ -316,12 +325,10 @@ struct ZigZag
...
-
-        const float_X min_1 = ::min(i_1, i_2) + float_X(1.0);
-        const float_X max_1 = ::max(i_1, i_2);
-        const float_X max_2 = ::max(max_1, (x_1 + x_2) / float_X(2.));
-        const float_X x_relayPoint = ::min(min_1, max_2);
-        return x_relayPoint;
+        /* paper version:
+         *   i_1 == i_2 ? (x_1 + x_2) / float_X(2.0) ::max(i_1, i_2);
+         */
+        return i_1 == i_2 ? x_2 : ::max(i_1, i_2);
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: is that [std::max](http://en.cppreference.com/w/cpp/algorithm/max) or a libPMacc `max`?

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 5]
```diff
@@ -316,12 +325,10 @@ struct ZigZag
...
     calc_relayPoint(const float_X i_1, const float_X i_2, const float_X x_1, const float_X x_2) const
     {
-
-        const float_X min_1 = ::min(i_1, i_2) + float_X(1.0);
-        const float_X max_1 = ::max(i_1, i_2);
-        const float_X max_2 = ::max(max_1, (x_1 + x_2) / float_X(2.));
-        const float_X x_relayPoint = ::min(min_1, max_2);
-        return x_relayPoint;
+        /* paper version:
+         *   i_1 == i_2 ? (x_1 + x_2) / float_X(2.0) ::max(i_1, i_2);
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: missing `:` in else branch

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp [hunk 6]
```diff
@@ -287,7 +290,8 @@ struct ZigZag
             /* this loop is only needed for 2D, we need a flux in z direction */
             for (uint32_t d = simDim; d < 3; ++d)
             {
-                flux[d] = float_X(0.5) *  charge * velocity[d] * volume_reci;
+                /* in 2D the full flux for the z direction is given to the virtual particle zero */
+                flux[d] = (l == float_X(0) ? charge * velocity[d] * volume_reci : float_X(0.0));
```
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: why don't you compare with the `parId` which is a `const int`?

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp (hunk 4 above)
Comment (psychocoderHPC) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: Currently libPMacc has no `max` implementation for `float` and `double` therefor the default cuda max/min is used.
The libPMacc min/max is coming in a separate pull request.

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp (hunk 6 above)
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: I don't see why depositing `half` in `flux[d]` should be the same as depositing it fully. that does not look equivalent.

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp (hunk 1 above)
Comment (psychocoderHPC) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: it is called "relay point" in the corresponding papers

File: src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp (hunk 6 above)
Comment (ax3l) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: ah you avoid the double deposition here, I see...

Comment (psychocoderHPC) [inline src/picongpu/include/fields/currentDeposition/ZigZag/ZigZag.hpp]: `parId == l` therefore both variables can used, but might be it is better to use the integer `l`

**Comment by psychocoderHPC** [comment]
**speedup test case 1:**

- 3D KHI electrons/ions
- global ~67*10^6 particles on 1 GPUs (k20)
- 64x64x64 cells
- SuperCellSize 8x8x4
- 2000 time steps

speedup normalized to Esirkepov

|  | shape     | Esirkepov | ZigZag-dev | ZigZag-currentPullRequ |
| --- | ---------- |-----:| ----:|----:|
| time [s] | TSC  | 3280 |  2158 | 1458 |
| speedup | TSC  | 1.0 |  1.5 |  2.2 |

**speedup test case 2:**

- 3D KHI electrons/positrons
- global ~67*10^6 particles on 1 GPUs (k20)
- 64x64x64 cells
- SuperCellSize 8x8x4
- 2000 time steps

speedup normalized to Esirkepov

|  | shape     | Esirkepov | ZigZag-dev | ZigZag-currentPullRequ |
| --- | ---------- |-----:| ----:|----:|
| time [s] | TSC  | 3799 |  2479 | 1778 |
| speedup | TSC  | 1.0 |  1.5 |  2.1 |

**Comment by psychocoderHPC** [comment]
heating 2D KHI positron/electron with x-drift (Esirkepov vs. current pull request)

16 GPUs k20
2000 steps

2d-x_heating

**Comment by PrometheusPi** [comment]
@psychocoderHPC You changed 4 lines of code and reduced simulation time by 30%! Kudos!

**Comment by bussmann** [comment]
Is that a new record?

**Comment by bussmann** [comment]
P.S.: Very well done!

**Comment by ax3l** [comment]
Thank you for the improvement, this is marvelous! :rocket:

**Pull request merged**